### PR TITLE
[Enhancement](compaction) reduce VMergeIterator copy block

### PR DESCRIPTION
**Pull request by BiteTheDDDDt**
# Proposed changes

This pr change make VMergeIterator support return row reference to instead copy a full block.

before:
```
output_version=[2-11], 
current_max_version=11, 
segments=50, 
input_row_num=30000000, 
output_row_num=30000000. 
elapsed time=53.7632s. 
cumulative_compaction_policy=SIZE_BASED, 
compact_row_per_second=558002
```
图片

after:
```
output_version=[2-11], 
current_max_version=11, 
segments=50, 
input_row_num=30000000, 
output_row_num=30000000. 
elapsed time=47.3122s. 
cumulative_compaction_policy=SIZE_BASED, 
compact_row_per_second=634086
```
图片

test case:
```sql
CREATE TABLE IF NOT EXISTS `customer` (
  `c_custkey` int(11) NOT NULL COMMENT "",
  `c_name` varchar(26) NOT NULL COMMENT "",
  `c_address` varchar(41) NOT NULL COMMENT "",
  `c_city` varchar(11) NOT NULL COMMENT "",
  `c_nation` varchar(16) NOT NULL COMMENT "",
  `c_region` varchar(13) NOT NULL COMMENT "",
  `c_phone` varchar(16) NOT NULL COMMENT "",
  `c_mktsegment` varchar(11) NOT NULL COMMENT ""
) ENGINE=OLAP
DUPLICATE KEY(`c_custkey`)
COMMENT "OLAP"
DISTRIBUTED BY HASH(`c_custkey`) BUCKETS 1
PROPERTIES (
"replication_num" = "1",
"colocate_with" = "groupa2",
"in_memory" = "false",
"storage_format" = "DEFAULT"
);

insert into customer select * from ssb.customer;
insert into customer select * from ssb.customer;
insert into customer select * from ssb.customer;
insert into customer select * from ssb.customer;
insert into customer select * from ssb.customer;
insert into customer select * from ssb.customer;
insert into customer select * from ssb.customer;
insert into customer select * from ssb.customer;
insert into customer select * from ssb.customer;
insert into customer select * from ssb.customer;
```

## Problem summary

Describe your changes.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [ ] No
    - [ ] I don't know
2. Has unit tests been added:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [ ] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [ ] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/olap/rowset/beta_rowset_reader.cpp [hunk 1]
```diff
@@ -313,6 +314,27 @@ Status BetaRowsetReader::next_block(vectorized::Block* block) {
...
+                if (s.is_end_of_file()) {
+                    return Status::OLAPInternalError(OLAP_ERR_DATA_EOF);
+                } else {
+                    LOG(WARNING) << "failed to read next block: " << s.to_string();
+                    return Status::OLAPInternalError(OLAP_ERR_ROWSET_READ_FAILED);
+                }
+            }
+        } while (block_view->empty());
+    } else {
+        return Status::NotSupported("block view only support enable_storage_vectorization");
```
Comment (Gabriel39) [inline be/src/olap/rowset/beta_rowset_reader.cpp:332]: If user set `enable_storage_vectorization` to false, he/she can never execute sql until turning this to `true` and restart cluster? I think we should make this more friendly, for example, if `enable_storage_vectorization` is false, use `next_batch` to do compaction or query execution

File: be/src/vec/core/block.h [hunk 2]
```diff
@@ -497,5 +497,24 @@ class MutableBlock {
     }
 };
 
+struct IteratorRowRef {
+    std::shared_ptr<Block> block;
+    int16_t row_pos;
```
Comment (HappenLee) [inline be/src/vec/core/block.h:502]: maybe here should use `int32_t`, it do not change the struct size.

File: be/src/olap/rowset/beta_rowset_reader.cpp (hunk 1 above)
Comment (BiteTheDDDDt) [reply be/src/olap/rowset/beta_rowset_reader.cpp:342]: use ```cpp get_data_by_ref = get_data_by_ref && _rs_reader->support_return_data_by_ref() &&
                       config::enable_storage_vectorization``` to control it now.

**Review by HappenLee — APPROVED** [review]
LGTM

**Pull request merged**